### PR TITLE
Add ESLint to satellite along with anti trojan plugin

### PR DESCRIPTION
**Pull request by dhillonks**
## Description
Added ESLint as a part of https://github.com/Seneca-CDOT/telescope/issues/2560 and https://github.com/Seneca-CDOT/telescope/issues/2533

## Changes:
- [X] Added and configured ESLint
- [X] Fixed linting errors
- [X] Configued ESLint with `jest`
- [X] Added [anti trojan source plugin](https://github.com/lirantal/eslint-plugin-anti-trojan-source)
- [X] Modified the pre-commit hook through husky to also run eslint

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,9 +7,13 @@
     "test:watch": "jest --watch",
     "test": "jest -c jest.config.js",
     "coverage": "jest -c jest.config.js --collect-coverage",
+    "eslint": "eslint --config .eslintrc.js \"**/*.{ts,js}\"",
```
Comment (humphd) [inline package.json:10]: We don't have an `.ts` in Satellite, so this can be `.js` only.

File: package.json [hunk 2]
```diff
@@ -7,9 +7,13 @@
...
+    "eslint-fix": "eslint --config .eslintrc.js \"**/*.{ts,js}\" --fix",
```
Comment (humphd) [inline package.json:11]: Same here

File: test.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-/* global describe, test, beforeEach, afterEach, expect */
+// global describe, test, beforeEach, afterEach, expect
```
Comment (humphd) [inline test.js:1]: I don't think these are necessary if you use the jest plugin

**Review by humphd — CHANGES_REQUESTED** [review]

File: test.js (hunk 3 above)
Comment (dhillonks) [reply test.js]: ESLint was earlier giving no-redeclare errors earlier: `'describe' is already defined as a built-in global variable    no-redeclare`. While switching to a single line comment seemed to resolve that, I'm not really sure if that comment is needed anymore so I think removing the comment altogether is a better approach. Please let me know it needs to be reverted

File: package.json (hunk 2 above)
Comment (dhillonks) [reply package.json]: Done!

File: package.json (hunk 1 above)
Comment (dhillonks) [reply package.json]: Done!

File: src/middleware.js [hunk 4]
```diff
@@ -15,6 +15,7 @@ function isAuthenticated() {
 }
 
 // Determine whether the authorization options passed to isAuthorized are valid.
+//eslint-disable-next-line no-unused-vars
```
Comment (humphd) [inline src/middleware.js:18]: What is it complaining about here?  Can we fix the issue instead?

File: jest.config.js [hunk 5]
```diff
@@ -13,7 +13,6 @@ module.exports = {
   bail: 1,
   verbose: true,
   testPathIgnorePatterns: ['/node_modules/'],
-  testEnvironment: 'node',
```
Comment (humphd) [inline jest.config.js:16]: Why this change?

File: .eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+     * Halt if a trojan source attack is found
+     * https://github.com/lirantal/eslint-plugin-anti-trojan-source
+     */
+    'anti-trojan-source/no-bidi': 'error',
+  },
+  overrides: [
+    {
+      files: ['**/test.js'],
+      rules: {
+        'no-unused-vars': 'off',
```
Comment (humphd) [inline .eslintrc.js:24]: Why do we need this?

**Review by humphd — CHANGES_REQUESTED** [review]

File: jest.config.js (hunk 5 above)
Comment (dhillonks) [reply jest.config.js:16]: This line was duplicated in the config file, present at both line 12 and 16.

File: src/middleware.js (hunk 4 above)
Comment (dhillonks) [reply src/middleware.js:18]: There is this function which is unused, causing eslint to throw `no-unused-vars`:
```
function validateAuthorizationOptions(options = {}) { ... }
```
So, I added the ignore statement. Now while, adding it to the exports would fix this problem, I wasn't sure if it should be exposed. Please let me know if that is more ideal:
```
module.exports.validateAuthorizationOptions= validateAuthorizationOptions;
```

File: .eslintrc.js (hunk 6 above)
Comment (dhillonks) [reply .eslintrc.js:24]: In the `test.js` file, there are no-unused-vars errors for variables at places like:
```js
((req, user) => { ... })
```
Even though the variables in such instances are technically unused, I felt that they do hold meaning. I looked into the options for the [no-unused-vars](https://eslint.org/docs/rules/no-unused-vars) rule but there's no suitable option that would avoid such cases in `test.js`. So, rather than adding `//eslint-disable-next-line` for each such error, I added that override for the `test.js` file

File: src/middleware.js (hunk 4 above)
Comment (humphd) [reply src/middleware.js:18]: This is a good example of where it's better to listen to eslint than ignore it.  In the past, we used to allow passing an Object to `isAuthorized()`, and we would validate it, see https://github.com/Seneca-CDOT/satellite/pull/8/files#diff-9a117d52a8a2cb3e5f498575bdfdea4dc32418180c09029454fbaf7f95ebe5a1R42.  However, it looks like we dropped that, and now require you to pass a function instead.  As a result, the right fix here is to delete this entire funcction, because it's not being used anymore.  We don't want to keep code that isn't being used.

File: .eslintrc.js (hunk 6 above)
Comment (humphd) [reply .eslintrc.js:24]: Let's disable this globally in the test.js file via a comment at the top vs. doing it for the whole project.

**Review by humphd — CHANGES_REQUESTED** [review]

File: .eslintrc.js (hunk 6 above)
Comment (dhillonks) [reply .eslintrc.js]: Done!

File: src/middleware.js (hunk 4 above)
Comment (dhillonks) [reply src/middleware.js]: Didn't know about that, thanks for the insight. Removed with the latest changes.

**Review by humphd — APPROVED** [review]

**Comment by humphd** [comment]
@dhillonks can you do another PR that builds on this to add the `lint` check to our CI/CD workflows? https://github.com/Seneca-CDOT/satellite/blob/master/.github/workflows

**Comment by dhillonks** [comment]
> @dhillonks can you do another PR that builds on this to add the `lint` check to our CI/CD workflows? https://github.com/Seneca-CDOT/satellite/blob/master/.github/workflows

on it!
